framework/src/media/OutputHandler.cpp

```diff
@@ -232,39 +251,28 @@ void *OutputHandler::workerMain(void *arg)
 		}
 
 		auto size = stream->mBufferReader->sizeOfData();
-		if ((size >= stream->mStreamBuffer->getThreshold()) ||
-			(size > 0 && stream->mIsFlushing)) {
-			auto buf = new unsigned char[size];
-			size_t wlen = 0;
-			while (wlen < size) {
-				size_t temp = stream->mBufferReader->read(buf + wlen, size - wlen);
-				if (temp == 0) {
-					// Can not read data, error occurred.
-					meddbg("StreamBufferReader::read failed!\n");
-					break;
-				}
-				wlen += temp;
-			}
-
-			if (worker->write(buf, size) < 0) {
-				// Error occurred, stop outputting
-				stream->mBufferWriter->setEndOfStream();
-				delete buf;
-				break;
+		if (stream->mIsFlushing) {
+			while (size > 0 && stream->writeToSource(size)) {
+				size = stream->mBufferReader->sizeOfData();
 			}
-			delete buf;
+			std::unique_lock<std::mutex> lock(stream->mFlushMutex);
+			stream->mIsFlushing = false;
+			stream->mFlushCondv.notify_one();
+		} else if (size >= stream->mStreamBuffer->getThreshold()) {
+			stream->writeToSource(size);
 		}
 	}
 
+	medvdbg("OutputHandler exit\n");
 	return NULL;
 }
 
 void OutputHandler::sleepWorker()
 {
 	size_t spaces = mBufferWriter->sizeOfSpace();
 	std::unique_lock<std::mutex> lock(mMutex);
-	// In case of flushing or overrun, DO NOT sleep worker.
-	if (mIsWorkerAlive && !mIsFlushing && (spaces > 0)) {
+	// In case of overrun, DO NOT sleep worker.
+	if (mIsWorkerAlive && (spaces > 0)) {
```

Comment:
Why do you remove `flushing` check?
Please think about below case.
```c
void OutputHandler::flush()
{
	...
	wakenWorker();
	////// If worker is alive and just sleeped here!  it may not be wakened and flush() is blocked... /////
	if (mIsWorkerAlive) {
		std::unique_lock<std::mutex> lock(mFlushMutex);
		mFlushCondv.wait(lock);  // worker may be sleeping, will not notify.
	}
	...
}
```

Reply:
Thanks for the review.
I fixed like below
```
void OutputHandler::flush()
{
    if (mIsWorkerAlive) {
        std::unique_lock<std::mutex> lock(mFlushMutex);
        mIsFlushing = true;
        wakenWorker(); // If worker wakes with mIsFlushing == true, It will stopped at std::unique_lock<std::mutex> lock(mFlushMutex) at worker
        mFlushCondv.wait(lock);
    }
}
```

Reply:
Looks good, nice work! :)